compiler-rt/src/crt0/allocator.cairo

```diff
@@ -0,0 +1,413 @@
+                // larger region can be allocated.
+                self.next_allocation += bytes_left_in_current_cell;
+                self.space_in_current_cell = false;
+            }
+        }
+
+        // Here, the next allocation pointer must be at the beginning of a new cell.
+        let current_allocation = self.next_allocation;
+        self.next_allocation += byte_count;
+
+        if byte_count % SIZEOF_CELL != 0 {
+            // Make a note we're not fully utilizing the last cell, so we may want to use it for
+            // the next allocation.
+            self.space_in_current_cell = true;
+        }
+
+        // Allocate cells in the mapping
+        let cells_needed = utils::cells_count_from_bytes(current_allocation, byte_count);
+        let mut cells_allocated: u128 = 0;
+        while cells_allocated < cells_needed {
+            let key: felt252 = (current_allocation + cells_allocated * SIZEOF_CELL).into();
+            const DEFAULT_MEMORY_VALUE: felt252 = 0;
+            self.allocated_addresses.insert(key, DEFAULT_MEMORY_VALUE);
+            cells_allocated += 1;
+        };
+
+        current_allocation
+    }
+
+    fn load(ref self: AllocatorState, address: Address, byte_count: ByteCount) -> Array<u8> {
+        let cells_needed = utils::cells_count_from_bytes(address, byte_count);
+        let offset_in_first_cell: Address = address % SIZEOF_CELL;
+        let cell_base_addr = utils::cells_base_address(address);
+        let mut cells_read: u128 = 0;
+        let mut bytes_to_read = byte_count;
+
+        // Buffer to store bytes of retrieved memory cells.
+        let mut output = array![];
+
+        while cells_read < cells_needed && bytes_to_read > 0 {
+            let key = (cell_base_addr + cells_read * SIZEOF_CELL);
+            let cell_value = self.allocated_addresses.get(key.into());
+
+            // If this is the first cell and there's an offset, we need to skip some bytes from the
+            // beginning of the cell. Otherwise we read the full cell from the beginning.
+            let (bytes_to_read_from_cell, address_in_cell) = if cells_read == 0
+                && offset_in_first_cell > 0 {
+                (min(bytes_to_read, SIZEOF_CELL - offset_in_first_cell), offset_in_first_cell)
+            } else {
+                const BEGINNING_OF_CELL: Address = 0;
+                (min(bytes_to_read, SIZEOF_CELL), BEGINNING_OF_CELL)
+            };
+
+            // Collect bytes from cell to the buffer.
+            for byte in utils::cell_to_buffer(cell_value)
+                .span()
+                .slice(
+                    // These conversions are safe, these values cannot exceed SIZEOF_CELL.
+                    address_in_cell.try_into().expect('load address exceeds cell size'),
+                    bytes_to_read_from_cell.try_into().expect('bytes to load exceed cell size'),
+                ) {
+                output.append(*byte);
```

Comment:
This is a question more than anything: are we able to do something akin to a `collect()` to avoid the intermediate 'reallocations' every time we append?